### ISSUE TITLE
Fix alt text doesn't work

### ISSUE THREAD

**Issue by jwbowles** [ucfopen/UDOIT#191]
I noticed that the fixAltText function in Ufixit.php is returning the img tag without the new alt text added properly. After investigating further it appears that
`$remove_attr = preg_replace("/ data-api-endpoint.+?>/", "", $fixed_img);
        $fixed_img = $remove_attr;
`
strips the data-api-endpoint tag as well as any text following that tag. This happens to include the alt tag because it is appended to the end of the new HTML string. This just recently started occurring without any modification to the code on our part.

**Comment by jwbowles** [comment]
Here is a fix I developed to address this issue:
`public function fixAltText($error_html, $new_content, $submitting_again = false)
    {
        $this->dom->loadHTML('<?xml encoding="utf-8" ?>' . $error_html);

        $imgs = $this->dom->getElementsByTagName('img');
        $fixed_img = null;

        foreach ($imgs as $img) {
            $img->setAttribute('alt', $new_content);
            $img->removeAttribute('data-api-endpoint');
            $img->removeAttribute('data-api-returntype'); 
            $fixed_img = $this->dom->saveHTML($img);
        }

        $fixed_img = trim($fixed_img,">");

        return $fixed_img;
    }`

This will remove data-api-endpoint and data-api-returntype attributes and then trim the trailing ">" from the tag so that the additional attributes Canvas requires will be appended to the end of the image tag.

**Comment by bagofarms** [comment]
Sorry i didn't get to review this issues sooner.  Thank you so much for providing the code.  I know we're backed-up on pull requests right now, but if you could submit one, that would be great!

**Comment by cooperfellows** [comment]
I've been looking a lot at this issue the past two days, hoping to have a discussion about the best way forward here. @jwbowles is right, the current system doesn't work, as the alt tag is removed by the regex as pointed out.

I started with his proposed fix from PR #192, but found that none of the text was replaced. Through some other trial and error ([you can see that history here](https://github.com/ucfopen/UDOIT/compare/master...cooperfellows:UDOIT_issue191)) I managed to get it working correctly. I cleaned up all the debug commits and the only change that we really need to make is [seen here](https://github.com/ucfopen/UDOIT/compare/master...cooperfellows:UDOIT_issue191_clean).

However, I am wondering why we are trying to remove those attributes in the first place? I understand not wanting to introduce extra code back into a page (or syllabus in this case), but this is not the way to do it.

Those data-* tags will just be re-introduced anytime that page's content is updated. This is due to the way Canvas returns the code that makes up a page. For example, if you have two images on one page, both missing their alt tags. The first time you fix an image, the data-* attribute will be removed from the first image BUT the second image (or any other images) will have those attributes ADDED because that's how the content is provided from the Canvas API ([example image here](https://www.screencast.com/t/qf7dR9tSjgR)). Then, when fixing the second image (or any other problem on that page) the first image, the one we just fixed and removed the data-* attributes from, will have those data-* attributes re-introduced by Canvas, and since we only try to strip them when fixing the individual image tag, they will simply be added back to the course content.

So, I think there are really two things to think about:
1. Ensure the fixImgAlt process works (ignoring the existence or introduction of the data-* tags)
    - To move this along, I have submitted a new PR
2. Decide how we want to handle these additional tags that Canvas adds
    - Just ignore them, and have them be added when we update a page
    - Or build some sort of cleanup process that removes any of these data-* tags introduced by canvas
       - This is much more involved, as we would need to define the set of data-* tags that Canvas adds and strip only those from content without removing data-* tags added by content authors.

**Comment by jwbowles** [comment]
Nice work @cooperfellows! My vote would be to ignore the data attributes if possible because the solution i presented is dependent on specific attributes. If Instructure ever adds another attribute we would need to add the new attribute removal step making this not very sustainable. With that said, in some of my past testing I noticed if the data-* attributes were not removed then the attributes were duplicated. I am currently unable to replicate that behavior, but will continue to pinpoint it. In addition I had some strange results when the data-* attributes were removed by the user through the html editor before running UDOIT and attempting to fix the alt attribute.

**Comment by bagofarms** [comment]
> However, I am wondering why we are trying to remove those attributes in the first place? I understand not wanting to introduce extra code back into a page (or syllabus in this case), but this is not the way to do it.

In the current version of UDOIT, we are stripping out the data-api-endpoint tags from both the [original HTML](https://github.com/ucfopen/UDOIT/blob/master/public/process.php#L155) (the one still containing the error) and the [corrected HTML](https://github.com/ucfopen/UDOIT/blob/master/lib/Ufixit.php#L120) (the one with the ALT text added).

We're doing this because the process of loading the HTML into a DOM object and saving it makes the results inconsistent from the raw HTML we get back from Canvas.  Removing these data attributes from the error_html allows us to [search for the exact string in the HTML content](https://github.com/ucfopen/UDOIT/blob/master/lib/Ufixit.php#L474) we get back from the API so that we can replace it with the corrected string.  We remove the data attributes from the corrected string so that Canvas can add them in again.  We didn't want to risk uploading a malformed data attribute because of all the steps we go through.

I'm sure there's a better way of doing this, but this is what's working for us.  I think @cooperfellows is on the right track with PR #223 , but you both make a good point that we should take some time to re-evaluate this procedure to see if it is even needed.
